lib/execution_segment.go

```diff
@@ -394,13 +395,151 @@ func (ess ExecutionSegmentSequence) String() string {
 	return strings.Join(result, ",")
 }
 
+// lowest common denominator
+// https://en.wikipedia.org/wiki/Least_common_multiple#Using_the_greatest_common_divisor
+func (ess ExecutionSegmentSequence) lcd() int64 {
+	var acc = ess[0].length.Denom().Int64()
+	var n int64
+	for _, seg := range ess[1:] {
+		n = seg.length.Denom().Int64()
+		if acc == n || acc%n == 0 { // short circuit
+			continue
+		}
+		acc = acc * n / gcd(acc, n)
+	}
+
+	return acc
+}
+
+// Greatest common divisor
+// https://en.wikipedia.org/wiki/Euclidean_algorithm
+func gcd(a, b int64) int64 {
+	for a != b {
+		if a > b {
+			a -= b
+		} else {
+			b -= a
+		}
+	}
+	return a
+}
+
 // GetStripedOffsets returns everything that you need in order to execute only
 // the iterations that belong to the supplied segment...
 //
 // TODO: add a more detailed algorithm description
-func (ess ExecutionSegmentSequence) GetStripedOffsets(segment *ExecutionSegment) (int, []int, error) {
-	start := 0
-	offsets := []int{}
-	// TODO: basically https://docs.google.com/spreadsheets/d/1V_ivN2xuaMJIgOf1HkpOw1ex8QOhxp960itGGiRrNzo/edit
-	return start, offsets, fmt.Errorf("not implemented")
+// TODO: basically https://docs.google.com/spreadsheets/d/1V_ivN2xuaMJIgOf1HkpOw1ex8QOhxp960itGGiRrNzo/edit
+func (ess *ExecutionSegmentSequence) GetStripedOffsets(segment *ExecutionSegment) (int64, []int64, int64, error) {
+	if segment == nil || segment.length.Cmp(oneRat) == 0 {
+		return 0, []int64{1}, 1, nil
+	}
+
+	// we will copy the sequnce to this in order to sort it :)
+	var copyESS ExecutionSegmentSequence
+	// Here we fix the problem with having no sequence
+	// No filling up is required as the algorithm will accommodate for it
+	// through just going through the iterations that need to be in the values will fill up
+	// this has the consequence that if this is ran without sequence,
+	// but with segments: 0:1/3 and 1/3:2/3 it will get the same results instead
+	// of 1/3:2/3 to get start=1 and offset={3} it will get as 0:1/3 will start=0 and offsets={3}
+	// if the above behaviour is desired this will definitely need to be outside of this function.
```

Comment:
This comment block is a bit difficult to understand (IMO). Particularly "the algorithm will accomodate for it through just going through the iterations that need to be in the values will fill up", it seems like there's a typo or grammar issue there. In general it could use more periods :)

Reply:
or commas :stuck_out_tongue_winking_eye: